drivers/network/dd/e1000/hardware.c

```diff
@@ -696,15 +742,35 @@ NICGetPermanentMacAddress(
     USHORT AddrWord;
     UINT n;
 
+    ULONG Rah;
+    ULONG Ral;
+
     NDIS_DbgPrint(MAX_TRACE, ("Called.\n"));
 
-    /* Should we read from RAL/RAH first? */
-    for (n = 0; n < (IEEE_802_ADDR_LENGTH / 2); ++n)
-    {
-        if (!E1000ReadEeprom(Adapter, (UCHAR)n, &AddrWord))
-            return NDIS_STATUS_FAILURE;
-        Adapter->PermanentMacAddress[n * 2 + 0] = AddrWord & 0xff;
-        Adapter->PermanentMacAddress[n * 2 + 1] = (AddrWord >> 8) & 0xff;
+    // First try to get the MAC via Ral/Rah. The first is read from NVM after a soft reset
+    // (which occurs in ndis.c just before calling this function)
+    E1000ReadUlong(Adapter, E1000_REG_RAH, &Rah);
+    if (Rah & E1000_RAH_AV) {
+        NDIS_DbgPrint(MIN_TRACE, ("MAC from RAH/RAL seems to be useable...\n"));
+        Rah &= 0xffff;
+        E1000ReadUlong(Adapter, E1000_REG_RAL, &Ral);
+
+        *(ULONG *)Adapter->PermanentMacAddress = Ral;
```

Comment:
I'd prefer bit shifting here instead of the casts, but won't insist if you or someone else feels strongly.

Reply:
I just copied it from some lines below ;-)